### DISCUSSION TITLE
How many threads Node creates per core when using worker threads?

### DISCUSSION THREAD

**Discussion by crealityeducation** [nodejs/help#4817] [Q&A]
I'm running node js worker threads on MacBook Air M1. There are 8 core. And using PISCINA to create threads. I'm getting 12 threads. Are those for one core or on all 8 core. How do I know if main thread is blocked/being used in worker threads when running workers for all threads?

When running on all threads, other requests are getting blocked. I want to use all cores and full capacity of threads without blocking main thread.

Also, when I see the CPU utilisation when worker threads are running, its not crossing above 30%. It looks like threads are spawned on single core.

**Comment by mhdawson** [comment]
@crealityeducation I think this is likely a better question for the piscina repo - https://github.com/piscinajs/piscina as that is what I believe will be creating the threads.

**Reply by crealityeducation** [reply]
Hello Mark, thanks for the response. Sure, I'll raise one on PISCINA as well, but if I dont use PISCINA, and use the node worker module to create worker thread, would NODEJS create thread on only one CPU Core or use all the CORE's available? I'm curious to know. Accordingly I'll update my implementation. I really need performance here. And I see the CPU utilization is not crossing 25% while in worker threads mode. It looks like I'm underutilizing resources.

**Reply by mhdawson** [reply]
By default Node.js has a single event loop were almost all work takes place, a few other support threads are created but realisticly its best to assume assume that all work will be contrained to the single thread used by the event loop.

To use additional threads you need to use worker threads. Each thread will get it's own event loop and can consume the resources of a thread.

**Reply by mhdawson** [reply]
Also here is some general guidance that we've put together where I work in terms of threading/scaling and multi-threading - https://github.com/nodeshift/nodejs-reference-architecture/blob/main/docs/functional-components/scaling-multi-threading.md
